torch/csrc/api/include/torch/data/ctf/ctf_parser.h

```diff
@@ -0,0 +1,931 @@
+#endif
+
+    // Decides whether this is a new example or an existing one
+    if (previous_sequence_id_ != sequence_id && sequence_id != LONG_MAX) {
+      is_new = true;
+    }
+
+    previous_sequence_id_ = sequence_id;
+    has_initial_sequence_id_ = true;
+    return is_new;
+  }
+
+  bool get_input_stream(
+      const CTFSequenceID& sequence_id,
+      bool& is_new_sequence) {
+    // Create a new sequence with input_streams_info pre-allocated
+    if (is_new_sequence) {
+      is_new_sequence = false;
+
+#ifdef CTF_DEBUG
+      dataset_->sequences_id.emplace_back(sequence_id);
+#endif
+
+      // New sequence to be appended to dataset_->sequences
+      CTFSequenceData sequence;
+
+      for (auto const& stream : dataset_->input_streams_info) {
+        CTFInputStreamID input_stream_id =
+            dataset_->input_streams_map[stream.name];
+        if (stream.storage == CTFDataStorage::Dense) {
+          // TODO: Performance consideration: CNTK knows the number of samples
+          // in the sequence, allowing accurate memory reservation (index built
+          // during init)
+          sequence.emplace_back(
+              std::make_shared<CTFDenseInputStreamData<DataType>>(
+                  input_stream_id, stream.dimension));
+        } else {
+          sequence.emplace_back(
+              std::make_shared<CTFSparseInputStreamData<DataType>>(
+                  input_stream_id, stream.dimension));
+        }
+      }
+
+      dataset_->sequences.emplace_back(sequence);
+    }
+
+    // Reads the Input Stream name and lookup its input stream reference
+    CTFInputStreamID input_stream_id;
+    if (!get_input_stream_name(input_stream_id)) {
+      return false;
+    }
+    const CTFInputStreamInformation& input_stream =
+        dataset_->input_streams_info[input_stream_id];
+
+    // Appends all values to the input stream
+    if (!get_input_stream(input_stream)) {
+      return false;
+    }
+
+    // TODO: Check actual number of values records of the stream
+    return true;
+  }
+
+  // Parses input name from buffer and returns both CTFInputStreamInformation
+  // reference and true if the input name belongs to an existing Input Stream
+  bool get_input_stream_name(CTFInputStreamID& input_stream_id) {
+#ifdef CTF_DEBUG
+    // For logging purposes
+    size_t initial_pos = reader_->get_position();
+#endif
+    // idx will be used to iterate through scratch_ for local string parsing
+    size_t idx = 0;
+
+    // CTF Name must start with a '|'
+    if (!is_name_prefix(reader_->peek_char())) {
+#ifdef CTF_DEBUG
+      std::cout << "Not a CTF Name at position " << initial_pos << std::endl;
+#endif
+      return false;
+    }
+
+    // Discard | and get all consecutive digits and alpha characters
+    char c = reader_->get_char();
+    while (is_digit(reader_->peek_char()) || is_alpha(reader_->peek_char())) {
+      c = reader_->get_char();
+      scratch_[idx++] = c;
```

Comment:
Do we need to keep a copy in scratch_ here? It looks like scratch_ actually stores data very similar to the buffer_ in reader_. 

To avoid extra memory usage and improve performance, I think we can keep track of the start and end position for the stream name, and convert the name directly from buffer_ in reader_. 

For example, if buffer_ contains a substring "feature" starting at index i1 and ending at index i2, we can get the name "feature" out by calling 
std::string name = std::string(buffer_.data() + i1, buffer_data() + i2) without copying to scratch_.